cesium/internal/domain/index.go

```diff
@@ -78,20 +102,57 @@ func (idx *index) overlap(tr telem.TimeRange) bool {
 	return overlap
 }
 
-func (idx *index) update(ctx context.Context, p pointer) (err error) {
+func (idx *index) update(ctx context.Context, p pointer, persist bool) error {
 	_, span := idx.T.Bench(ctx, "update")
-	idx.mu.RLock()
+	idx.mu.Lock()
+
+	defer span.End()
+
 	if len(idx.mu.pointers) == 0 {
-		idx.mu.RUnlock()
-		return span.EndWith(RangeNotFound)
+		// This should be inconceivable since update would not be called with no pointers.
+		idx.mu.Unlock()
+		idx.L.DPanic(RangeNotFound.Error())
+		return span.Error(RangeNotFound)
 	}
 	lastI := len(idx.mu.pointers) - 1
 	updateAt := lastI
 	if p.Start != idx.mu.pointers[lastI].Start {
 		updateAt, _ = idx.unprotectedSearch(p.Start.SpanRange(0))
 	}
-	idx.mu.RUnlock()
-	return span.EndWith(idx.updateAt(ctx, updateAt, p))
+
+	ptrs := idx.mu.pointers
+	oldP := ptrs[updateAt]
+	if oldP.Start != p.Start {
+		// This should never happen since update would only be called via commit, and
+		// commit should find the same pointer the writer has been writing to, which
+		// must have the same Start timestamp. Unhandled race conditions might cause the
+		// database to reach this inconceivable state.
+		idx.mu.Unlock()
+		idx.L.DPanic(RangeNotFound.Error())
+		return span.Error(RangeNotFound)
+	}
+	overlapsWithNext := updateAt != len(ptrs)-1 && ptrs[updateAt+1].OverlapsWith(p.TimeRange)
+	overlapsWithPrev := updateAt != 0 && ptrs[updateAt-1].OverlapsWith(p.TimeRange)
+	if overlapsWithPrev || overlapsWithNext {
+		idx.mu.Unlock()
+		return span.Error(ErrDomainOverlap)
+	} else {
+		idx.mu.pointers[updateAt] = p
+	}
+
+	idx.persistHead = min(idx.persistHead, updateAt)
+
+	if persist {
+		update := indexUpdate{afterIndex: idx.persistHead}
+		idx.persistHead = len(idx.mu.pointers) - 1
+		idx.mu.Unlock()
+		idx.Observer.Notify(ctx, update)
```

Comment:
is section of code looks eerily similar to the code on line 74. Can we consolidate?